### ISSUE TITLE
Interpretation of the intra-modal contrastive (IMC) loss

### ISSUE THREAD

**Issue by ytaek-oh** [lezhang7/Enhance-FineGrained#15]
Dear authors,
Thank you for sharing the codes and checkpoint from your wonderful CVPR work! 
The motivation for treating hard negative texts, which finally derives IMC and CMR loss terms, was really cool to me!

However, while reading the paper, the formulation of the IMC loss was confusing for me. 
To my understanding, its motivation is to make a further distinction between the (original) caption and hard negative captions in the text embedding space, which is clear to me, but I cannot connect this to the exact design of equation (3) in the paper. 
Is there any reference or intuition behind it?

I also checked its implementation, which does not seem to align with the equation. 
It rather tries to maximize the similarity between the (original) caption and the first hard negative caption by the contrastive loss, if I am not mistaken.

```python
### src/training/data.py
class HardNegative_Collate:

    def __call__(self, batch):
        img = torch.stack([example[0] for example in batch])
        ture_caption = torch.stack([example[1][0] for example in batch])
        hard_negative = torch.cat([example[1][1:] for example in batch])
        text = torch.cat([ture_caption, hard_negative])
        valid_caption_mask = torch.stack([example[2] for example in batch])
        return img, text, valid_caption_mask

### src/open_clip/loss.py
# gt_text_features = text_features[:image_features.shape[0]]
# da_text_features = text_features[image_features.shape[0]:]  <- contains only hard negatives?
# embedding_matrix = logit_scale * gt_text_features @ da_text_features.T  #(batch_size,4*batch_size)
def get_imc_loss(self, gt_logits_per_image: torch.Tensor, embedding_matrix: torch.Tensor):
    """
    gt_logits_per_image: standard clip similarity matrix, diag is true gt similarity value : shape [batch_size,5xbatch_size]
    embedding_matrix: extra similarity matrix served as denominator in clip loss
    """

    logtis_matrix = embedding_matrix
    labels = torch.zeros(logtis_matrix.shape[0], device=logtis_matrix.device, dtype=torch.long)
    imc_loss = F.cross_entropy(logtis_matrix, labels)
    return imc_loss

```

In my humble opinion, prepending an additional column corresponding to the positive caption representation to the `embedding_matrix` is necessary for the contrastive loss to work desirably.

Could you share your insights or point out anything that I'm missing?

Best regards,

**Comment by lezhang7** [comment]
Hi,

Equation 3 maximizes the similarity between true and generated hard negatives, which is equivalent to minimizing $\sum \log{\sum \exp^{s(T, T_k)}}$.

During implementation, we found that the swap words hard negative, essentially $T_{rel}$, is crucial for learning compositional understanding. It is key to keep the threshold for relation $Th_{rel}$ increasing, as shown in Figure 5b. To accelerate this, it is important to add a term $-S(T, T_{rel})$ to the CMR loss to guide the representation optimization in a certain direction.

The implementation is simplified and combined as follows:

$$
-\sum \log \frac{S(T, T_{rel})}{\sum S(T, T_k)} = \sum \left(\log{\sum \exp^{s(T, T_k)}} - S(T, T_{rel})\right)
$$

Here, the numerator $S(T, T_{rel})$ is for the CMR loss, not the IMC loss.

**Comment by ytaek-oh** [comment]
Thank you for the detailed reply!
I missed the addition of the term $-S(T, T_{rel})$ to the IMC loss (eqn 3).

Based on my understanding, minimizing the IMC loss (eqn 3) is equivalent to minimizing the similarities between the correct and generated negative captions, and the added term stabilizes the training for the CMR loss, which are combined in the implementation of the IMC loss.

The formulation and motivation behind it is now clear to me.
